File: week2/reverse_words_in_a_string.py
"""
Jing 3/20, 2014
http://oj.leetcode.com/problems/reverse-words-in-a-string/
"""
class Solution:
    # @param s, a string
    # @return a string
    def reverseWords(self, s):
        return ' '.join(s.split()[::-1]) # This is the fastest
        #return ' '.join(reversed(s.split())) #second best
        #this one also passed the test but not very good
        #return ' '.join([word for word in reversed(s.split())])

File: week2/reverse_words_in_a_string.cpp
/* Accepted
 * By Yao
 * Date: March 28, 2014
 * From http://oj.leetcode.com/problems/reverse-words-in-a-string/
 */
#include <iostream>
#include <string>

using namespace std;

string reverseWords(const string &s){
    char word[100];
    string res;
    int i=0;
    int j=0;
    int status =0;
    while(s[i]!='\0'){
       if(s[i]!=' ' &&  s[i] !='\t' && s[i]!='\n') {
           status = 1;
           word[j++] = s[i];
       }
       if(status && (s[i] == ' ' || s[i] == '\t'|| s[i]=='\n')){
           status = 0;
           if(!res.length())
               res  = string(word, j) + res;
           else
               res = string(word, j) + ' ' + res;
           j = 0;
       }
       i++;
    }
    if(j && res.length())
        res = string(word, j) + ' ' + res;
    if(j && !res.length())
        res = string(word, j);
    
    return res;
}

int main(){
    string s;
    getline(cin,s);

    cout<<reverseWords(s)<<endl;
}
